Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -0,0 +1,212 @@
+        AZStd::string fullTopic = ROS2Names::GetNamespacedName(GetNamespace(), publisherConfig.m_topic);
+        m_pointCloudPublisher = ros2Node->create_publisher<sensor_msgs::msg::PointCloud2>(fullTopic.data(), publisherConfig.GetQoS());
+
+        SetPhysicsScene();
+        if (m_sensorConfiguration.m_visualise)
+        {
+            auto* entityScene = AZ::RPI::Scene::GetSceneForEntityId(GetEntityId());
+            m_drawQueue = AZ::RPI::AuxGeomFeatureProcessorInterface::GetDrawQueueForScene(entityScene);
+        }
+        m_lidarRaycaster.SetAddPointsMaxRange(m_lidarParameters.m_addPointsAtMax);
+        ROS2SensorComponent::Activate();
+    }
+
+    void ROS2LidarSensorComponent::Deactivate()
+    {
+        ROS2SensorComponent::Deactivate();
+        m_pointCloudPublisher.reset();
+    }
+
+    void ROS2LidarSensorComponent::FrequencyTick()
+    {
+        float distance = m_lidarParameters.m_maxRange;
+        auto entityTransform = GetEntity()->FindComponent<AzFramework::TransformComponent>(); // TODO - go through ROS2Frame
+        const auto directions =
+            LidarTemplateUtils::PopulateRayDirections(m_lidarParameters, entityTransform->GetWorldTM().GetEulerRadians());
+        AZ::Vector3 start = entityTransform->GetWorldTM().GetTranslation();
+        start.SetZ(start.GetZ());
+
+        auto worldToLidarTransform = entityTransform->GetWorldTM();
+        worldToLidarTransform.Invert();
+
+        m_lastScanResults =
+            m_lidarRaycaster.PerformRaycast(start, directions, worldToLidarTransform, distance, m_ignoreLayer, m_ignoredLayerIndex);
+        if (m_lastScanResults.empty())
+        {
+            AZ_TracePrintf("Lidar Sensor Component", "No results from raycast\n");
+            return;
+        }
+
+        if (m_sensorConfiguration.m_visualise)
+        { // Store points for visualisation purposes, in global frame
+            auto localToWorldTM = entityTransform->GetWorldTM();
+
+            // TODO - improve performance
+            m_visualisationPoints = m_lastScanResults;
+            for (AZ::Vector3& point : m_visualisationPoints)
+            {
+                point = localToWorldTM.TransformPoint(point);
+            }
+        }
+
+        auto* ros2Frame = Utils::GetGameOrEditorComponent<ROS2FrameComponent>(GetEntity());
+        auto message = sensor_msgs::msg::PointCloud2();
+        message.header.frame_id = ros2Frame->GetFrameID().data();
+        message.header.stamp = ROS2Interface::Get()->GetROSTimestamp();
+        message.height = 1;
+        message.width = m_lastScanResults.size();
+        message.point_step = sizeof(AZ::Vector3); // TODO - Point Fields can be custom
+        message.row_step = message.width * message.point_step;
+
+        // TODO - a list of supported fields should be returned by lidar implementation
+        std::vector<std::string> point_field_names = { "x", "y", "z" };
```

Comment:
1. The TODO should be made into a Github Issue
2. The std::vector of std::string can be replaced with `char*[]` which would avoid any heap allocations.
```suggestion
        // TODO - a list of supported fields should be returned by lidar implementation
        char* point_field_names[] = { "x", "y", "z" };
```

Reply:
Used `AZStd::array<const char*, 3> point_field_names = { "x", "y", "z" };`  later `.size()` is used.

Reply:
Actually, {x, y, z} will be used almost always, but extra fields will often be present (such as "intensity" or "i", "layer" or "ring", "return", "distance", "source", ....)